architecture/design/pessimistic-locking-functional-spec.md

```diff
@@ -0,0 +1,661 @@
+## [Functional Spec] YSQL Pessimistic Locking
+
+
+## 1 Introduction
+
+To understand pessimistic locking behaviour in Pg, it is important to take note of some points about row-level locking and its interaction with DMLs.
+
+<span style="text-decoration:underline;">PostgreSQL supports 4 types of row-level locks -</span>
+
+
+
+1. Exclusive modes (i.e., only 1 txn can hold such a lock at any time) -
+    1. `FOR UPDATE` - takes exclusive lock on the whole row (blocks the shared locks as well)
+    2. `FOR NO KEY UPDATE` - takes exclusive lock on row but allows share lock on pk i.e., `FOR KEY SHARE`
+2. Share access modes
+    3. `FOR SHARE` - shared lock on full row i.e., blocks all exclusive locks
+    4. `FOR KEY SHARE` - shared lock on key i.e., allows `FOR NO KEY UPDATE`
+
+<span style="text-decoration:underline;">PostgreSQL has two notions of conflict -</span>
+
+
+
+1. **Lock-modification conflict: **(modification is a change in tuple via a DML) _(refer example 2)._
+    1. `FOR UPDATE`: doesn’t allow any concurrent modification to the locked tuple.
+    2. `FOR NO KEY UPDATE`: doesn’t allow any concurrent modification to the locked tuple.
+    3. `FOR SHARE`: doesn’t allow any concurrent modification to the locked tuple.
+    4. `FOR KEY SHARE`: allow only modification to non-pk cols
+
+Lock-modification conflict only happens when the period from lock acquire to lock release intersects/ overlaps with the read to commit time of the transaction performing the modification.
+
+Lock-modification conflicts result in serialization errors. Note that the lock might be issued before or after the modification (refer example 2i and 2ii).
+
+Note that in READ COMMITTED isolation, when a lock-modification conflict occurs, Pg does some post-processing (called “READ COMMITTED update checking steps”) instead of throwing a 40001.
+
+
+
+2. **Lock-lock conflict:** this conflict only refers to one lock blocking another i.e., just waiting behaviour. It **doesn’t** refer to a serialization error. In other words, two conflicting locks can only result in blocking and not in a serialization error (or “write” conflict as we roughly call it). Also note that these locks can be taken explicitly using a `SELECT FOR `or implicitly using a DML. So there can be 4 cases - implicit-implicit, implicit-explicit, explicit-implicit, explicit-explicit lock-lock conflicts.
+
+    _(refer example 1 for explicit lock-explicit lock conflict)._
+
+
+DMLs implicitly take the following row-level locks:
+
+
```

Comment:
Everywhere: more than one blank line gets ignored by the renderer

Reply:
makes sense